### PR TITLE
Time range abilities

### LINKED ISSUE #579: Cancancan: "can't iterate from ActiveSupport::TimeWithZone"
I think currently we cannot define time period ability like [this](https://github.com/Coursemology/coursemology2/blob/master/app/models/course/assessment/assessment_ability.rb#L56). 

Otherwise you will get an error of `can't iterate from ActiveSupport::TimeWithZone` when `load_and_authorize_resource`

This is because `TimeWithZone` is not iterable, code like `(1.day.ago..1.day.from_now).include? Time.now` will fail. See [here] (https://github.com/CanCanCommunity/cancancan/blob/develop/lib/cancan/rule.rb#L143)

Also `Time.min` or `Time.max` are invalid inputs for Postgres, Sql generated by `XXX.accessible_by(current_ability)` will fail when there is a `Time.min` or `Time.max` defined in ability.

I think we might need to find a new way of defining the time abilities, any suggestions ?

### PR DESCRIPTION
**Pull request by allenwq**
This will fix #579 

I submitted a fix to cancancan: https://github.com/CanCanCommunity/cancancan/pull/276

### CODE REVIEW

File: lib/extensions/date_time_helpers.rb [hunk 1]
```diff
@@ -1,11 +1,12 @@
 module Extensions::DateTimeHelpers
   module TimeClassMethods
     def min
-      at(-Float::MAX)
+      # 200 BC, almost the smallest time that postgres can support
+      at(-2**36)
```
Comment (lowjoel) [inline lib/extensions/date_time_helpers.rb]: Maybe 1/1/0000 is good enough...

File: spec/features/course/assessment_management_spec.rb [hunk 2]
```diff
@@ -70,7 +70,8 @@
       let(:user) { create(:course_user, :approved, course: course).user }
       scenario 'I can see assessments' do
         category = course.assessment_categories.first
-        assessment = create(:assessment, course: course, tab: category.tabs.first)
+        assessment = create(:assessment, course: course, tab: category.tabs.first,
+                                         start_at: 1.day.from_now)
```
Comment (lowjoel) [inline spec/features/course/assessment_management_spec.rb]: why change this?

**Comment by lowjoel** [comment]
Can you link the cancancan PR also?

Also, why did the page break, but the feature specs did not trigger?

**Comment by allenwq** [comment]
@lowjoel I added a link in the description. 

Actually feature tests fails on my local for few weeks already, I'm also not sure why travis one does not fail. 

So I intentionally set the assessment time to a future time, I think that will let it fail.

**Comment by lowjoel** [comment]
Go back to the original cancancan release and check if the specs break. I do not think that this is the cause.

**Comment by allenwq** [comment]
As I said, specs always break on my local. 

Just reverted, let's see how. I remember I checked this before submitting the issue.

**Comment by lowjoel** [comment]
Look carefully at which spec broke, it's not likely at all that it would pass on Linux and Windows, but fail on Mac OS

**Comment by allenwq** [comment]
@lowjoel You are right. Things becomes interesting now, it was because ruby version.

`(Time.min..Time.max).include? Time.now`  This expression gives exception on 2.2.1 but returns true on 2.2.3

**Comment by lowjoel** [comment]
you can keep the rest of the fixes

File: lib/extensions/date_time_helpers.rb (hunk 1 above)
Comment (allenwq) [inline lib/extensions/date_time_helpers.rb]: `Time.utc(0)` ?  then how about the max time ? `Time.utc(5000)` maybe ?

**Pull request merged**